docs/2020-06-17-Checking-Implication.md

```diff
@@ -0,0 +1,161 @@
+- [All-Path](2019-03-28-All-Path-Reachability-Proofs.md)
+
+## Problem
+
+Consider a reachability claim,
+
+```
+φ(X) → ∘ ∃ Y. ψ(X, Y)
+```
+
+where `∘` denotes the appropriate one-path or all-path modality
+and `X` and `Y` denote _distinct_ families of variables.
+We will assume that `φ(X)` and `ψ(X, Y)` are function-like patterns (matching at most one value)
+because this is true in typical use.
+
+The reachability proof algorithm will unroll the modality, giving
+
+```
+φ(X) → ∃ Y. ψ(X, Y) ∨ • ∘ ∃ Y. ψ(X, Y)
+```
+
+If the direct implication holds,
+
+```
+φ(X) → ∃ Y. ψ(X, Y)
+```
+
+then the claim is proven.
+Otherwise, the algorithm continues with
+
+```
+φ(X) → • ∘ ∃ Y. ψ(X, Y)
+```
+
+by rewriting `φ(X)`.
+Here, were are concerned with checking the direct implication,
+
+```
+φ(X) → ∃ Y. ψ(X, Y)
+```
+
+or equivalently, checking that
+
+```
+φ(X) ∧ ¬ ∃ Y. ψ(X, Y)
+```
+
+is unsatisfiable.
+
+## Solution
+
+Checking that the formula
+
+```
+φ(X) ∧ ¬ ∃ Y. ψ(X, Y)
+```
+
+is unsatisfiable is equivalent to checking
+
+```
+⌈φ(X) ∧ ¬ ∃ Y. ψ(X, Y)⌉.
+```
+
+Because we assume that `φ(X)` is a function-like pattern,
+we can move the negation outside of `⌈ ⌉`:
+
+```
+⌈φ(X) ∧ ¬ ∃ Y. ψ(X, Y)⌉
+===
+¬ ⌈φ(X) ∧ ∃ Y. ψ(X, Y)⌉ ∧ ⌈φ(X)⌉.
+```
+
+By alpha-equivalence, we can also move the quantifier outside of `⌈ ⌉`:
+
+```
+(¬ ⌈φ(X) ∧ ∃ Y. ψ(X, Y)⌉) ∧ ⌈φ(X)⌉
+===
+(¬ ∃ Y. ⌈φ(X) ∧ ψ(X, Y)⌉) ∧ ⌈φ(X)⌉.
+```
+
+## Optimization
+
+We can optimize this solution the typical case where
+
+```
+φ(X) = t₁(X) ∧ P₁(X)
+ψ(X, Y) = t₂(X, Y) ∧ P₂(X, Y)
+```
+
+where `t₁(X)` and `t₂(X, Y)` are term-like patterns
+and `P₁(X)` and `P₂(X, Y)` are predicates:
+
+```
+\not(∃ Y. ⌈φ(X) ∧ ψ(X, Y)⌉) ∧ ⌈φ(X)⌉
+===
+\not(∃ Y. ⌈t₁(X) ∧ t₂(X, Y)⌉ ∧ P₁(X) ∧ P₂(X, Y)) ∧ ⌈t₁(X)⌉ ∧ P₁(X)
+===
+\not((∃ Y. ⌈t₁(X) ∧ t₂(X, Y)⌉ ∧ P₂(X, Y)) ∧ P₁(X)) ∧ ⌈t₁(X)⌉ ∧ P₁(X)
+===
+\or(\not(∃ Y. ⌈t₁(X) ∧ t₂(X, Y)⌉ ∧ P₂(X, Y)), ¬ P₁(X)) ∧ ⌈t₁(X)⌉ ∧ P₁(X)
+===
+\or(\not(∃ Y. ⌈t₁(X) ∧ t₂(X, Y)⌉ ∧ P₂(X, Y)) ∧ ⌈t₁(X)⌉ ∧ P₁(X), ¬ P₁(X) ∧ ⌈t₁(X)⌉ ∧ P₁(X))
+===
+\not(∃ Y. ⌈t₁(X) ∧ t₂(X, Y)⌉ ∧ P₂(X, Y)) ∧ ⌈t₁(X)⌉ ∧ P₁(X)
+```
+
+The sub-term `⌈t₁(X) ∧ t₂(X, Y)⌉` is a unification problem with an efficient internal implementation.
+The unification problem should be solved first because
+
+```
+⌈t₁(X) ∧ t₂(X, Y)⌉ → ⌈t₁(X)⌉
+```
+
+so that if unification succeeds, there is no need to explicitly consider `⌈t₁(X)⌉`.
+If unification fails, then the negative conjunct is `\top` so that the implication is satisified when `⌈t₁(X)⌉ ∧ P₁(X)` is unsatisfiable.
+Now, the implication is checked in two cases:
+after a configuration is rewritten (in which case the configuration may be rewritten again)
+or after a configuration can no longer be rewritten (a remainder pattern).
+In the case of a rewritten configuration, when unification fails, there is no need to check the rest of the implication pattern because the condition `⌈t₁(X)⌉ ∧ P₁(X)` is ensured by subsequent rewrites.
```

Comment:
This is true because the backend stops the proof and returns an error whenever the z3 solver can't figure out whether a predicate is satisfiable or unsatisfiable, right?

Reply:
It's true because the pattern we are rewriting is `t₁ ∧ P₁`, and if we add the new condition then we would be rewriting `t₁ ∧ P₁ ∧ ⌈t₁⌉ ∧ P₁`, which represents the same pattern. I should rewrite this section to be clearer about that.